src/lsp/cobol_parser/expect/gen_expect.ml

```diff
@@ -0,0 +1,421 @@
+
+let () =
+  Arg.parse
+    Arg.[
+      ("--external-tokens", Set_string external_tokens,
+       "<module> Import token type definition from <module>");
+    ]
+    anon usage_msg
+
+let cmlyname = match !cmlyname with
+  | None | Some "" -> Fmt.epr "%s@." usage_msg; exit 1
+  | Some s -> s
+
+(* --- UTILS --- *)
+
+let inv f a b = f b a
+
+let pp_brackets pp =
+  Fmt.(any "[" ++ pp ++ any "]")
+
+let pp_pair pp_a pp_b = fun ppf (a, b) ->
+    Fmt.pf ppf "(%a,%a)" pp_a a pp_b b
+
+let pp_list pp = Fmt.list ~sep:(Fmt.any ";") pp
+
+let rec pp_nel pp ppf = function
+  | NEL.One v -> Fmt.pf ppf "One %a" pp v
+  | v::nel -> Fmt.pf ppf "%a::" pp v; pp_nel pp ppf nel
+
+let pp_match_cases ppf pp_key pp_value default l =
+  List.iter (fun (keys, value) ->
+    Fmt.pf ppf "  | %a -> %a\n"
+      Fmt.(list ~sep:(any " | ") pp_key) keys
+      pp_value value) l;
+  Fmt.pf ppf "  | _ -> %s\n" default
+
+(* --- *)
+
+include MenhirSdk.Cmly_read.Read (struct let filename = cmlyname end)
+
+type completion_entry =
+  | K of terminal NEL.t
+  | Custom of string
+(* [@@deriving ord] *)
+
+let pp_completion_entry: completion_entry Fmt.t = fun ppf -> function
+  | K list -> Fmt.pf ppf "K (%a)" (pp_nel Print.terminal) list
+  | Custom custom_type -> Fmt.pf ppf "%s" custom_type
+
+let is_valid_for_comp: terminal -> bool = fun term ->
+  Terminal.kind term == `REGULAR &&
+  Terminal.attributes term |>
+  List.exists (fun attrib ->
+    Attribute.has_label "keyword" attrib ||
+    Attribute.has_label "completion" attrib ||
+    Attribute.has_label "keyword.combined" attrib)
+
+let terminal_filter_map: terminal -> completion_entry option = fun term ->
+  if is_valid_for_comp term then Some (K (NEL.One term)) else None
+
+let nonterminal_filter_map: nonterminal -> completion_entry option = fun nonterm ->
+  Nonterminal.attributes nonterm |>
+  List.find_opt (Attribute.has_label "completion") |>
+  Option.map Attribute.payload |>
+  inv Option.bind (fun s -> Some (Custom s))
+
+let pp_xsymbol_of_nt ppf nonterminal =
+  Fmt.pf ppf "X(N N_%s)" (Nonterminal.mangled_name nonterminal)
+
+(* --- *)
+
+module SuperSet (OrderedType : Set.OrderedType) = struct
+  include Set.Make(OrderedType)
+
+  let add_option elt t =
+    match elt with
+    | None -> t
+    | Some elt -> add elt t
+
+  let add_list l t = add_seq (List.to_seq l) t
+
+  let from_list_map ?(init=empty) ~add f l =
+    List.fold_left (fun acc value -> add (f value) acc) init l
+end
+
+module SymbolSet = SuperSet(struct
+  type t = symbol
+  let compare = Symbol.compare
+end)
+
+module CompEntrySet = struct
+  include SuperSet(struct
+    type t = completion_entry
+    let compare entry1 entry2 =
+      match entry1, entry2 with
+        | Custom s1, Custom s2 -> String.compare s2 s1
+        | K nel1, K nel2 -> NEL.compare Terminal.compare nel1 nel2
+        | Custom _, K _ -> -1
+        | K _, Custom _ -> 1
+  end)
+end
+
+module Map5 (OrderedType : Map.OrderedType) = struct
```

Comment:
Why `5`? Also this is a `Make_map`

Reply:
because add_to_list comes from ocaml 5.1 standard lib

Reply:
You can add this as a comment near the function definition, but that's not worth adding in the functor name.